obspy/core/event/resourceid.py

```diff
@@ -0,0 +1,653 @@
+
+    # A dict for keeping track of parent/object specific references.
+    # weak_key, weak_key, weak_value
+    # {parent_key: {resource_key: object_id}}
+    _parent_id_tree = WeakKeyDictionary()
+
+    # Dict of list to keep track of the order in which resource_ids are
+    # assigned to object_ids.
+    # {resource_id: [object_id, object_id, ...]}
+    _id_order = WeakKeyDictionary()
+
+    # A weak value dict mapping object ids (not resource ids) to objects.
+    # {object_id: object}
+    _id_object_map = WeakValueDictionary()
+
+    # set default _ResourceKey attributes and object_id
+    _parent_key = _ResourceKeyDescriptor('_parent_key')
+    _resource_key = _ResourceKeyDescriptor('_resource_key')
+    _object_id = None
+
+    def __init__(self, id=None, prefix="smi:local", referred_object=None,
+                 parent=None):
+        # Create a resource id if None is given and possibly use a prefix.
+        if id is None:
+            self.fixed = False
+            self._prefix = prefix
+            self._uuid = str(uuid4())
+        elif isinstance(id, ResourceIdentifier):
+            self.__dict__.update(id.__dict__)
+            return
+        else:
+            self.fixed = True
+            self.id = id
+        # get resource and parent _ResourceKey singletons
+        self._resource_key = self.id
+        self._parent_key = parent
+        # set referred object if one was provided.
+        if referred_object is not None:
+            self.set_referred_object(referred_object)
+
+    def get_referred_object(self):
+        """
+        Returns the object associated with the resource identifier.
+
+        This works as long as at least one ResourceIdentifier with the same
+        ID as this instance has an associated object. If not, this method will
+        return None.
+        """
+        try:
+            return ResourceIdentifier._id_object_map[self._object_id]
+        except KeyError:
+            return self._get_similar_referred_object()
+
+    def _get_similar_referred_object(self):
+        """
+        Find an object with the same resource id.
+
+        If the resource_identifier instance is not bound to a specific object,
+        or the bound object has been garbage collected, try to find another
+        object that was assigned the same resource_id.
+
+        The parent_id_tree will be scanned first to see if any object_ids
+        have been assigned to the same resource_id in the parent scope. If
+        not simply use the last object assigned the resource_id. If no object
+        is found return None.
+        """
+        # Warn if resource_id was bound but its object got gc'ed.
+        if self._object_id is not None:
+            msg = ("The object with identity of: %d no longer exists, "
+                   "returning the most recently created object with a"
+                   " resource id of: %s") % (self._object_id, self.id)
```

Comment:
This warning message is not fully correct as it would also be triggered if the object is found in the parent scope.

Reply:
fixed